### DISCUSSION TITLE
RFC: Control Flow Analysis

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#222] [RFC]
In this RFC we discuss about our learnings and findings on various CFG implementations across different JavaScript / TypeScript compilers.

Requirement:
* Graph data backed by a vec or a memory arena
* Can be used in the linter and also the minifier, which means graph construction has be separated out
* Can be visualized

Background:
* https://rustc-dev-guide.rust-lang.org/appendix/background.html#cfg

**Comment by Boshen** [comment]
## TypeScript

**Reply by Boshen** [reply]
This is in binder.ts

**Comment by Boshen** [comment]
## ESLint

* [code-path-analysis](https://github.com/eslint/eslint/tree/main/lib/linter/code-path-analysis)

**Comment by Boshen** [comment]
## Rome

* https://github.com/rome/tools/pull/2781
* https://github.com/rome/tools/pull/2763

**Reply by MichaReiser** [reply]
Relevant PRs:

* https://github.com/rome/tools/pull/2781
* https://github.com/rome/tools/pull/2763

And you can use the [playground](https://docs.rome.tools/playground/?indentStyle=space&quoteStyle=single&trailingComma=es5&code=ZgB1AG4AYwB0AGkAbwBuACAAdABlAHMAdAAoACkAIAB7AAoACgBpAGYAIAAoAGEAKQAgAHsACgAgACAAIAByAGUAdAB1AHIAbgAgADEAOwAgAAoAfQAKAAoAIAAgAHIAZQB0AHUAcgBuACAAMgA7AAoAfQA%3D) to inspect the control flow graph of any function.

**Comment by Boshen** [comment]
## Rust

* https://rustc-dev-guide.rust-lang.org/appendix/background.html#cfg
* https://rust-lang.github.io/rfcs/1211-mir.html

**Comment by Boshen** [comment]
## V8

[Turbofan IR](https://docs.google.com/presentation/d/1Z9iIHojKDrXvZ27gRX51UxHD-bKf1QcPzSijntpMJBM/edit#slide=id.g19134d40cb_0_35)
